src/ray/core_worker/core_worker.cc

```diff
@@ -2412,6 +2433,16 @@ bool CoreWorker::PinExistingReturnObject(const ObjectID &return_id,
   }
 }
 
+ObjectID CoreWorker::AllocateDynamicReturnId() {
+  const auto &task_spec = worker_context_.GetCurrentTask();
+  const auto return_id =
+      ObjectID::FromIndex(task_spec->TaskId(), worker_context_.GetNextPutIndex());
```

Comment:
At this point, we don't know how many objects will be returned, since this happens one at a time for each yield statement. I prefer to use `GetNextPutIndex` because it already has the semantics of incrementing a counter for each new object.

That reminds me, though, let me add a test for recovery where the task also calls `ray.put`.